govcd/service_account.go

```diff
@@ -0,0 +1,359 @@
+		version, err := client.GetVcdFullVersion()
+		if err == nil {
+			return nil, fmt.Errorf("minimum version for Service Accounts is 10.4.0 - Version detected: %s", version.Version)
+		}
+		// If we can't get the VCD version, we return API version info
+		return nil, fmt.Errorf("minimum API version for Service Accounts is 37.0 - Version detected: %s", client.APIVersion)
+	}
+
+	saParams := &types.ApiTokenParams{
+		ClientName:      name,
+		Scope:           scope,
+		SoftwareID:      softwareId,
+		SoftwareVersion: softwareVersion,
+		ClientURI:       clientUri,
+	}
+
+	newSaParams, err := vcdClient.RegisterToken(orgName, "37.0", saParams)
+	if err != nil {
+		return nil, fmt.Errorf("failed to register Service account: %s", err)
+	}
+
+	org, err := vcdClient.GetOrgByName(orgName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Org by name: %s", err)
+	}
+
+	serviceAccountID := "urn:vcloud:serviceAccount:" + newSaParams.ClientID
+	serviceAccount, err := org.GetServiceAccountById(serviceAccountID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Service account by ID: %s", err)
+	}
+
+	return serviceAccount, nil
+}
+
+func (sa *ServiceAccount) Update(saConfig *types.ServiceAccount) (*ServiceAccount, error) {
+	client := sa.org.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointServiceAccounts
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	saConfig.ID = sa.ServiceAccount.ID
+	saConfig.Name = sa.ServiceAccount.Name
+	saConfig.Status = sa.ServiceAccount.Status
+	urlRef, err := client.OpenApiBuildEndpoint(endpoint, saConfig.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	returnServiceAccount := &ServiceAccount{
+		ServiceAccount: &types.ServiceAccount{},
+		org:            sa.org,
+	}
+
+	err = client.OpenApiPutItem(apiVersion, urlRef, nil, saConfig, returnServiceAccount.ServiceAccount, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error updating Service Account: %s", err)
+	}
+
+	return returnServiceAccount, nil
+}
+
+// AuthorizeServiceAccount authorizes a service account and returns a DeviceID and UserCode which will be used while granting
+// the request, and sets the Service Account in `Requested` status
+func (sa *ServiceAccount) Authorize() error {
+	client := sa.org.client
+
+	uuid := extractUuid(sa.ServiceAccount.ID)
+	data := bytes.NewBufferString(
+		fmt.Sprintf("client_id=%s",
+			uuid,
+		))
+
+	resp, err := client.doTokenRequest(sa.ServiceAccount.Org.Name, "device_authorization", "37.0", "application/x-www-form-urlencoded", data)
+	if err != nil {
+		return err
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(body, &sa.authParams)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GrantServiceAccount Grants access to the Service Account and sets it in `Granted` status
+func (sa *ServiceAccount) Grant() error {
+	client := sa.org.client
```

Comment:
Do we need to check the service account current status?
If not, does the API return a clear error message if we do the wrong thing?

Reply:
Updated. Added a check if the userCode is set, if yes, then it's atleast authorized. If it's already granted it will say that it failed to grant the SA as the usercode is invalid.